### PR TITLE
Free memory to avoid OOM when resume optimizer states

### PR DESCRIPTION
**Pull request by BestJuly**
# What does this PR do ?
## Summary

When resuming training **with optimizer states loaded** (the default behavior), the first training iteration can OOM during the logging phase. After `train_step()` completes, PyTorch's CUDA caching allocator still holds freed activation memory. When `training_log()` subsequently triggers NCCL collectives (aux-loss all-reduce across PP/DP groups, timer barrier, etc.), the NCCL internal buffer allocations push GPU memory over the limit.

The fix calls `torch.cuda.empty_cache()` once on the first iteration, right before `training_log()`, to return cached-but-unused memory to the CUDA driver and make room for NCCL buffers.

- Only runs on the first iteration — no steady-state performance impact.
- `empty_cache()` does not affect any live tensors or training correctness.
- Adding `--no-load-optim` avoids the OOM (optimizer states are zero-initialized instead of loaded), confirming that the extra memory from resumed optimizer states is the root cause.

## Reproduction

The issue reproduces with a MoE proxy model using the distributed optimizer and expert parallelism. Reproduction:

```bash
torchrun --nproc_per_node=8 pretrain_gpt.py --distributed-timeout-minutes 220 --tensor-model-parallel-size 1 --pipeline-model-parallel-size 1 --expert-model-parallel-size 8 --context-parallel-size 1 --expert-tensor-parallel-size 1 --use-distributed-optimizer --no-create-attention-mask-in-dataloader --cross-entropy-loss-fusion --cross-entropy-fusion-impl te --enable-experimental --use-mcore-models --sequence-parallel --use-flash-attn --disable-bias-linear --micro-batch-size 1 --global-batch-size 32 --train-samples 268554688 --exit-duration-in-mins 230 --manual-gc --manual-gc-interval 5 --transformer-impl transformer_engine --tokenizer-type HuggingFaceTokenizer --tokenizer-model Qwen/Qwen3-235B-A22B --mock-data --split 99,1,0 --no-mmap-bin-files --num-workers 6 --untie-embeddings-and-output-weights --position-embedding-type rope --rotary-percent 1.0 --rotary-base 1000000 --rotary-seq-len-interpolation-factor 1 --normalization RMSNorm --swiglu --norm-epsilon 1e-06 --num-layers 10 --hidden-size 8192 --ffn-hidden-size 12288 --num-attention-heads 64 --kv-channels 128 --group-query-attention --num-query-groups 4 --qk-layernorm --seq-length 4096 --max-position-embeddings 4096 --make-vocab-size-divisible-by 1187 --attention-dropout 0.0 --hidden-dropout 0.0 --clip-grad 1.0 --weight-decay 0.1 --lr-decay-samples 584765624 --lr-warmup-samples 1536000 --lr-warmup-init 3.9e-7 --lr 3.9e-6 --min-lr 3.9e-7 --lr-decay-style cosine --adam-beta1 0.9 --adam-beta2 0.95 --num-experts 16 --moe-ffn-hidden-size 1536 --moe-router-load-balancing-type aux_loss --moe-router-topk 8 --moe-grouped-gemm --moe-aux-loss-coeff 1e-3 --moe-permute-fusion --moe-router-dtype fp32 --moe-router-fusion --eval-iters 32 --eval-interval 500 --auto-detect-ckpt-format --load ./results/debug_optim_load/checkpoints --save ./results/debug_optim_load/checkpoints --save-interval 10 --dist-ckpt-strictness log_all --init-method-std 0.02 --log-timers-to-tensorboard --log-memory-to-tensorboard --log-num-zeros-in-grad --log-params-norm --log-validation-ppl-to-tensorboard --log-throughput --log-interval 1 --logging-level 40 --tensorboard-dir ./results/debug_optim_load/tensorboard --bf16 --moe-token-dispatcher-type alltoal
```
After checkpointing saving and re-launch the script, without the PR it will OOM but works well by adding `--no-load-optim`. This PR helps to resume optimizer states without OOM error.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

**Comment by asolergi-nv** [comment]
We have #3558 which presents a similar issue. Should we investigate this further? In #3558 they are calling `torch.cuda.empty_cache()` earlier during checkpoint load.

**Comment by jaredcasper** [comment]
> We have #3558 which presents a similar issue. Should we investigate this further? In #3558 they are calling `torch.cuda.empty_cache()` earlier during checkpoint load.

It seems the excess memory does come from checkpoint load? (otherwise why would we only need to do this on the first iteration?). If that's the case I think #3558 makes more sense, making it clear it's releasing memory from checkpoint load. As is here it's a bit confusing why its only done on the first iteration.

**Comment by BestJuly** [comment]
Thanks @asolergi-nv and @jaredcasper for the info. I think the issue should be the same and I have verified the [PR#3558](https://github.com/NVIDIA/Megatron-LM/pull/3558https://github.com/NVIDIA/Megatron-LM/pull/3558) can also help with my case. Close this PR as that PR make it more clear for memory releasing.
